### ISSUE TITLE
How to resolve runtime error in `inst_eval` caused by hardcoding k_sample

### ISSUE THREAD

**Issue by NakedKoala** [mahmoodlab/CLAM#48]
```
 def inst_eval(self, A, h, classifier): 
        device=h.device
        if len(A.shape) == 1:
            A = A.view(1, -1)
        top_p_ids = torch.topk(A, self.k_sample)[1][-1]
        top_p = torch.index_select(h, dim=0, index=top_p_ids)
        top_n_ids = torch.topk(-A, self.k_sample, dim=1)[1][-1]
        top_n = torch.index_select(h, dim=0, index=top_n_ids)
        p_targets = self.create_positive_targets(self.k_sample, device)
        n_targets = self.create_negative_targets(self.k_sample, device)

        all_targets = torch.cat([p_targets, n_targets], dim=0)
        all_instances = torch.cat([top_p, top_n], dim=0)
        logits = classifier(all_instances)
        all_preds = torch.topk(logits, 1, dim = 1)[1].squeeze(1)
        instance_loss = self.instance_loss_fn(logits, all_targets)
        return instance_loss, all_preds, all_targets

```

In my data, there is large variance for # of instances per slide. Hard coding k_sample for this function does not work. 

i) Runtime error when # of instances < k_sample 
ii) when 2 * k_sample > # instances on current slide, then we will label some instance both positive and negative ??

I plan to use dynamic k_sample.  For each slide that contains N instances, I will draw N // 2 positive samples and N // 2 negative samples. 

Do you see any problem with this naive approach ? 

This code relies on sorting order to determine the classification "target". 

If I only have 2 instances and have very similar score, I am not sure the naive approach described above would make sense. ( In this case, we just create target arbitrarily )

**Comment by fedshyvana** [comment]
Yeah i think generally it might be reasonable to assume that the drawn top instances and bottom instances will be distinct if N >>>> k_sample. Just throwing ideas, but in your case instead of hard coding, it might be worth to try convert scores to percentiles, and drawing lets say the top 10 percentile to be your positive instances (if this is less than 1, then maybe just use the top 1), and the bottom 10 percentile to be your negative instances. You can cap it at lets say a maximum of 50 instances or something if you don't want the number of instances drawn from different slides to be too different .
